### ISSUE TITLE
Inconsistent custom emoji shortcode behaviour when lacking whitespace

### ISSUE THREAD

**Issue by undergroundmonorail** [mastodon/mastodon#7364]
When two custom emojis are inserted into a toot with no whitespace between them (e.g. `:glaceon::glaceon_shiny:` as opposed to `:glaceon: :glaceon_shiny:`), they will sometimes render as emojis and other times as just the shortcode. For example:

[Here's a toot where the first pair of emojis have no space between them, but they render as emojis](https://glaceon.social/@monorail/99973570350903999) (ironically, that toot is asking about if I can force them to render as emojis after a previous test didn't work that way).

[On the other hand, here's an entire thread where I tried to get it to happen again](https://glaceon.social/@monorail/99973593608628574), only succeeding when using the same text as the toot where it happened the first time.

Despite my best efforts, I can't identify anything about the original toot that would change how adjacent emojis would be rendered. Regardless of the cause, it's unintuitive and (at least imo) should probably be changed.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by nightpool** [comment]
This is weird and extremely edge-casey but i don't think the behavior is particularly unsound in any way—basically, we render emoji in two passes, one that determines *which* custom emoji are in a toot, and one that *replaces* those shortcodes with the emoji image itself. the second method uses a different technique then the first method, just a simple find/replace, so it occasionally will replace "extra" emoji that wouldn't have gotten replaced on their own.

This is obviously very weird and confusing behavior, but because of how obscure it is, and the fact that it's very hard to trigger accidentally, i dont know if it's worth the added complicated logic to fix it.

**Comment by undergroundmonorail** [comment]
I don't think it's particularly hard to trigger it accidentally, I ran into it trying to split an image into multiple emojis. Unless you include each emoji separately, [like this](https://glaceon.social/@monorail/99974338831287989), you'll only see the shortcodes.

**Comment by Gargron** [comment]
For one, actual emojis are unicode, not shortcodes, so spaces don't matter at all.

**Comment by undergroundmonorail** [comment]
I understand the technical reason why it works, but it makes sense to me to replicate that behaviour for custom emojis.

**Comment by nightpool** [comment]
@undergroundmonorail it's not just custom emoji—any shortcode processing needs the spaces. this has been mastodons behavior for a long time, so for example

`2600::5f86::100::68dc`

doesn't render as 

`2600::5f86:💯:68dc`

**Comment by undergroundmonorail** [comment]
I wasn't aware of that, thanks. Is there a particular reason that this is desirable behaviour?

**Comment by Gargron** [comment]
People really like IPv6 addresses :man_shrugging:

**Comment by undergroundmonorail** [comment]
Ohh, I hadn't considered IPv6 addresses. I'm not sure I know what they look like, actually. :P

**Comment by Gargron** [comment]
Like this:

>2600::5f86::100::68dc

**Comment by undergroundmonorail** [comment]
Oh. My bad :P

**Comment by nightpool** [comment]
yeah no worries. honestly i'm not sure if we even made the right call on
this one, but it's not very urgent, either way.

**Comment by cyberalien** [comment]
Maybe change parser to not parse set of sequences only if they all match `:[0-9a-f]+:`?

Or parse such sequences only if all of them can be parsed, such as `:100::100::100:`, which would clearly not be IPv6 address?

Those are simple solutions to make it work without affecting IPv6.

**Comment by vmstan** [comment]
This issue was filed against a now unsupported version of Mastodon. In an effort to clean up stale bugs, if this can be reliably reproduced on an updated version please open a new issue with updated logs/troubleshooting and we will take a look. Thank you.

**Comment by Neustradamus** [comment]
To follow this ticket.

**Comment by nikclayton** [comment]
Confirmed, still a problem.

This content:

> Here's a :blobcat:. Here are two more :blobcat::blobcat:.

renders all three shortcodes as emojis.

This content:

> Here's two emojis, :blobcat::blobcat:

doesn't render either of them (tested on mastodon.social running v4.5.0-nightly.2025-07-17).

I suspect (but have not verified) that this is not just a rendering issue. In the second example I don't believe Mastodon picked up that there's a custom emoji in the text at all, so when the status was federated no custom emojis were attached. So the bug effectively propogates to other servers that could display the emoji image but can't because it's not present.

**Comment by ThisIsMissEm** [comment]
The only reason that `Here's a :blobcat:. Here are two more :blobcat::blobcat:.` renders all three emojis is because of that first one. The `CustomEmoji.from_text` method extracts the first custom emoji, but fails for the second two.

However, clients are less strict about replacing custom emojis, it seems, so they just replace all occurrences of `:blobcat:` with that custom emoji.

Let's say you have two custom emojis `:blobcat:` and `:blobfoxdealwithit:` both from `hachyderm.io`, you can reproduce the emoji not being extracted with:
```ruby
CustomEmoji.from_text("Here's a :blobcat:. Here are two more :blobcat::blobfoxdealwithit:.", 'hachyderm.io')
```

It will only return you the `:blobcat:` custom emoji, and won't match the last two emojis.
